DOC-1782: API documentation for searchable menu buttons with new example (#2520)
tinymce/tinymce-docs: modules/ROOT/pages/custom-menu-toolbar-button.adoc

[cols="1,2,1,4",options="header"]
|===
|Name |Value |Requirement |Description

> |fetch |`+(success: (menu) => void, fetchContext) => void+` |required |Function that takes a callback and a `+fetchContext+` object. The callback function must be passed the list of options for the button's dropdown. The `+fetchContext+` object provides information which is useful for choosing which items should be passed to the callback. For more details on `fetchContext`, see xref:searchable-menu-buttons[Searchable Menu Buttons]

Author: I looked in various places, and we didn't seem to have a standard way of specifying the types of objects passed into functions, so because that detail is necessary to use this function, I've just put a small amount of detail in the table, and then linked to a lower section that has more information. The lower section still doesn't explicitly state the types, though. As I said, other examples (https://www.staging.tiny.cloud/docs/tinymce/6/dialog-configuration/) aren't particularly consistent on how they detail the information in objects passed through to functions. In some cases, the information doesn't seem to be present at all.

The admonition about this being only in >= 6.2 was also moved to the linked section, instead of putting even more content in this table (as was discussed).

Reviewer: Since this option is used in all menu buttons, not just the searchable menu button, I don't think referencing the searchable menu buttons makes this clear. I wonder if stating that there are examples of this function in both live demos.

Author: I wonder if the `Searchable Menu Buttons` part should just be called something closer to `FetchContext`. Let's clear this up with the team. This page differs from some of the others, which has made it hard to apply precedent and consistency.

Author: We'll address this later as part of a "Cleanup Searchable Menu docs page"

The `+fetch+` function is called when the toolbar menu button's menu is opened. It is a function that takes a callback and passes it an array of menu items to be rendered in the drop-down menu. This allows for asynchronous fetching of the menu items.

[[searchable-menu-buttons]]

> == Searchable Menu Buttons

Author: - [ ] inconsistent capitalisation

Author: ```suggestion
== Searchable menu buttons
```

> include::partial$misc/admon-requires-6.2v.adoc[]

Author: - [ ] ensure file is merged first (or add it to this PR)

> include::partial$misc/admon-requires-6.2v.adoc[]

Reviewer: This file exists within the footnotes PR, you should be able to copy it to this PR. This was done in the 6.1 release and no conflicts occur (as far as I can remember).

> include::partial$misc/admon-requires-6.2v.adoc[]

Author: Yeah, that was the intent. That someone else would merge it first. I have it locally, but didn't see the need to add it to the PR.

The button's menu can be configured to have an input field for searching, as well as its usual items. The presence of the input field is controlled by the `+search+` field in the xref:options[options]. The `+search+` field can be a `+boolean+` or an `+object+` containing a single optional string `placeholder`. By default, `+search+` is false. If `+search+` is not false, the button's menu will contain an input field with any specified placeholder text. As the user types into this field, the `+fetch+` function will be called with the text in the input field passed back as part of `+fetchContext+`. The `+fetch+` function is responsible for using this `+fetchContext+` to determine which items to pass to its `+success+` callback.

The `+fetchContext+` is an object containing a single string property: `+pattern+`. If the toolbar menu button has not configured `search` to be active, then the `+pattern+` string will be empty.

> == Searchable menu button example and explanation

Reviewer: ```suggestion
=== Searchable menu button example and explanation
```
This currently has the same heading level as `Searchable Menu Buttons`

Author: Yeah, that was intentional, because the other demo has this level. I think we probably just need to have a team discussion about how we want this file setup.

include::partial$misc/admon-requires-6.2v.adoc[]

The following is a simple toolbar menu button example, where searching has been configured:

> liveDemo::custom-toolbar-menu-button-search[tab="js"]

Author: Initially I had the Searchable Menu header section above the first demo, but it just became really hard to follow, so I moved the whole section below the first demo so that all the searchable menu content is in one place. I'm not tied to it. @brian-forte change however you see fit.

Author: Note: also this demo requires 6.2 (as you would expect)

> This example configures a toolbar menu button with the label `+My searchable button+` that opens the specified menu when clicked. The menu will contain a search input field because `+search+` is not `+false+`. The input field's https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#placeholder[placeholder attribute] will be `+What do you want to search for?+`.

Author: Heads-up: I've probably been inconsistent when using `blah` vs `+blah+`.

> This example configures a toolbar menu button with the label `+My searchable button+` that opens the specified menu when clicked. The menu will contain a search input field because `+search+` is not `+false+`. The input field's https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#placeholder[placeholder attribute] will be `+What do you want to search for?+`.

Reviewer 2: ```suggestion
This example configures a toolbar menu button with the label `+My searchable button+` that opens the specified menu when clicked. The menu will contain a search input field because `+search+` is not `+false+`. The input field's https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#placeholder[placeholder attribute] will be `+Type...+`.
```
